### PR TITLE
Developer tool: a decoder ring for stack traces

### PR DESCRIPTION
**Pull request by foldleft**
This relies on sourcemaps existing.

It should autodetect the bundle, so user just has to enter the filename, line and column.

Definitely needs polish, but for now I want to get the core functionality in place.

Since it's a tool, it's got a lot more of my own idiosyncratic style than usual, so, sorry about that.  `fold` is basically a `switch` statement for my datatypes.

This is deliberately written as bare js/html, so I didn't have to figure out an entire build system for this.  This does mean I've written raw element creation for React, rather than using JSX.

The broad design is to use RxJS to handle state changes, and React Hooks for rendering.  I use RxJS to create a stream of values for the result of a `fetch()`, consisting of progress values followed by a final successful chunk of data.  The idea is to feel snappy by providing user feedback as quickly as possible.

Land first.
Relates to https://github.com/vector-im/riot-web/issues/13359

### CODE REVIEW

File: res/decoder-ring/decoder.js [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+        const reader = readableStream.getReader();
+        const utf8Decoder = new TextDecoder("utf-8");
+        function readNextChunk() {
+            reader.read().then(({ done, value }) => {
+                if (done) {
+                    pendingSubject.complete();
+                    resultObservable.next(Success.of(buffer));
+                    return;
+                }
+                if (!value) {
```
Comment (foldleft) [inline res/decoder-ring/decoder.js:73]: debug code

File: res/decoder-ring/decoder.js [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+     * the number of bytes received with whatever pendingContext has been supplied.
+     * Finally, it will emit a Success containing the result.
+     * You'd use this on a Response.body.
+     */
+    function observeReadableStream(readableStream, pendingContext = {}) {
+        let bytesReceived = 0;
+        let buffer = "";
+        const pendingSubject = new rxjs.BehaviorSubject(Pending.of({ ...pendingContext, bytesReceived }));
+        const throttledPending = pendingSubject.pipe(rxjs.operators.throttleTime(100));
+        const resultObservable = new rxjs.Subject();
```
Comment (foldleft) [inline res/decoder-ring/decoder.js:63]: BehaviorSubject?

File: res/decoder-ring/decoder.js [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+                    pendingSubject.complete();
+                    resultObservable.next(Success.of(buffer));
+                    return;
+                }
+                if (!value) {
+                    console.log("empty chunk?", value);
+                    readNextChunk();
+                }
+                bytesReceived += value.length;
+                pendingSubject.next(Pending.of({bytesReceived, ...pendingContext}));
```
Comment (foldleft) [inline res/decoder-ring/decoder.js:78]: this doesn’t match the order above

File: res/decoder-ring/decoder.js [hunk 4]
```diff
@@ -117,7 +117,7 @@
             fetchStatus.fold({
             pending: ({ bytesReceived, length }) => {
                 if (!bytesReceived) {
-                    return e('span', { className: "spinner" }, "⦶");
+                    return e('span', { className: "spinner" }, "\u29b5");
```
Comment (bwindels) [inline res/decoder-ring/decoder.js:120]: I think `<meta charset="utf-8">` would also have fixed this btw, assuming you're encoding your files as utf8.

Comment (foldleft) [reply res/decoder-ring/decoder.js:120]: oh, whoops.

File: res/decoder-ring/datatypes.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
+'use strict';
+(function(global) {
+    class Optional {
```
Comment (turt2live) [inline res/decoder-ring/datatypes.js:3]: There's 3 different design patterns in these 3 lines :(

We don't need `use strict` in 2020, and the closure doesn't seem needed given we're using classes anyways.

Comment (foldleft) [reply res/decoder-ring/datatypes.js:3]: huh, old habits die hard I suppose!

**Comment by jryans** [comment]
What's the reason for adding this at the React SDK layer?

**Comment by foldleft** [comment]
I suppose it felt like it made semantic sense to me to do so, as it's next to `rageshake.js` then.

**Comment by foldleft** [comment]
moved to riot-web

**Pull request closed without merging**